### PR TITLE
feat: convert to pyproject.toml

### PR DESCRIPTION
**Pull request by MindTooth**
Goals:

* Use `pyproject.toml` as much as possible
* Check if `pip-tools` can be used for package management
* General cleanup of packaging

### CODE REVIEW

**Comment by MindTooth** [comment]
```
Run mypy.................................................................Failed
- hook id: mypy
- exit code: 1

src/vendor/sort/sort.py:24: error: Function is missing a type annotation  [no-untyped-def]
src/vendor/sort/sort.py:29: error: Function is missing a type annotation  [no-untyped-def]
src/vendor/sort/sort.py:52: error: Function is missing a type annotation  [no-untyped-def]
src/vendor/sort/sort.py:67: error: Function is missing a type annotation  [no-untyped-def]
src/vendor/sort/sort.py:97: error: Function is missing a type annotation  [no-untyped-def]
src/vendor/sort/sort.py:140: error: Function is missing a type annotation  [no-untyped-def]
src/vendor/sort/sort.py:150: error: Function is missing a return type annotation  [no-untyped-def]
src/vendor/sort/sort.py:164: error: Function is missing a return type annotation  [no-untyped-def]
src/vendor/sort/sort.py:171: error: Function is missing a type annotation  [no-untyped-def]
src/vendor/sort/sort.py:221: error: Function is missing a type annotation  [no-untyped-def]
src/vendor/sort/sort.py:233: error: Function is missing a type annotation  [no-untyped-def]
Found 11 errors in 1 file (checked 40 source files)
```

~~Does that mean mypy atm. can't read from pyproject.toml?~~  Installing the `toml` package did not help.  So this meaning it's been broken always?

**Comment by tomrtk** [comment]
> ~Does that mean mypy atm. can't read from pyproject.toml?~ Installing the `toml` package did not help. So this meaning it's been broken always?

SORT is not typed as it is vendored code, so it has been excluded here: https://github.com/tomrtk/fish-code/blob/3beb1082e48fe908ac8ebd1bb8da24b0095f09d7/.pre-commit-config.yaml#L6

edit: and in setup.cfg.

Unsure why it is type checked now in this pr.

**Comment by tomrtk** [comment]
I would suspect that mypy run by pre-commit do not read the correct mypy config, maybe a verbose output indicate what config is used?

File: pyproject.toml [hunk 1]
```diff
@@ -64,8 +53,48 @@ omit = [
...
+no_implicit_optional = true
+warn_redundant_casts = true
+plugins = "sqlalchemy.ext.mypy.plugin"
+
+[[tool.mypy.overrides]]
+module = "mypy-tests.*"
+disallow_untyped_defs = false
+
+[[tool.mypy.overrides]]
+module = "mypy-vendor.*"
```
Comment (tomrtk) [inline pyproject.toml]:
```suggestion
module = "vendor.*"
```
this should solve it I think

File: pyproject.toml [hunk 2]
```diff
@@ -64,8 +53,48 @@ omit = [
...
+
+[tool.mypy]
+check_untyped_defs = true
+disallow_untyped_defs = true
+no_implicit_optional = true
+warn_redundant_casts = true
+plugins = "sqlalchemy.ext.mypy.plugin"
+
+[[tool.mypy.overrides]]
+module = "mypy-tests.*"
```
Comment (tomrtk) [inline pyproject.toml]:
```suggestion
module = "tests.*"
```

File: pyproject.toml (hunk 1 above)
Comment (MindTooth) [reply pyproject.toml]: Yeah.  I knew somehting was a bit fishy with that one. 😉

**Comment by MindTooth** [comment]
Okay.  Now the pipeline succeeds, and the config is converted as much as possible (looking at you, flake8).

---

Next, how do we want to handle dependencies?  I tried toying around with `requirements.txt` as an export of `setup.cfg`, but, not sure if they made sense.  So, would it be possible to forgo `setup.cfg` all together and use `pyproject.toml` only and move flake8 config into its own file (`.flake8`)?  And then look at making the flow better (venv, packaging, etc.)?

**Comment by tomrtk** [comment]
> Next, how do we want to handle dependencies? I tried toying around with `requirements.txt` as an export of `setup.cfg`, but, not sure if they made sense. 

I think I would still define the top-level dependencies in `setup.cfg` or `pyproject.toml` (or `requirements.in`) pinned with a minimum version, and then use pip-tools to generate a pinned version of deps and sub-deps in a `requirements.txt` for a deploy-able-version.

> So, would it be possible to forgo `setup.cfg` all together and use `pyproject.toml` only and move flake8 config into its own file (`.flake8`)? 

Should be possible :+1:  

> And then look at making the flow better (venv, packaging, etc.)?

Not 100% sure what you would like to improve? do you think of dev. setup or ci?
Setup of dev setup I agree it could be simplified.  for ci as well there are probably improvements to be done.

**Comment by tomrtk** [comment]
> So, would it be possible to forgo `setup.cfg` all together and use `pyproject.toml` only and move flake8 config into its own file (`.flake8`)?

one think I remembered, at one point you could not install a package as editable with pip without a empty `setup.cfg/py`. so unless `setuptools` implementation for using `pyproject.toml` has matured, we may need to keep `setup.cfg` around.

**Comment by MindTooth** [comment]
Steps:

1. Move flake8 config into `.flake8`.
2. Migrate content of `setup.cfg` to `pyproject.toml`.
3. Verify that `pip install --editable` still works.
4. Update docs for setting up a virtual environment for testing and deployment.  Using `pip-tools` for generating a `requirements.txt` file from `pyproject.toml`.

Sounds about right?

**Comment by tomrtk** [comment]
> Sounds about right?

Yes, sounds right. I am wondering how to handle extras such as cpu, gpu...

**Comment by MindTooth** [comment]
Seeing newer Torch is being distributed differently, I think that the way going forward is just to install normally.  An pprovide an extra index if it needs to be overridden.

```
pip install --editable=. --extra-index-url https://download.pytorch.org/whl/cpu
```

**Comment by MindTooth** [comment]
Something with flake8 broke though. 🤣

**Comment by MindTooth** [comment]
> Something with flake8 broke though. 🤣

Fixed.

**Comment by MindTooth** [comment]
The `[project]` section needs a lot of love.  Some of the dynamic entires needs to be fixed.  Also, the packaging might now be completely broken.  The install command above works when running it inside a virtual environment, but I've not tried to install a package built by the new system.  Next on the agenda.

**Comment by MindTooth** [comment]
Can we with `pip-tools` perhaps remove a lot of the dependencies that are listed explicit, if they are already pulled implicitly by our hard requirements?  I noticed with `pip-compile`, it kinda functioned liked `package-lock.json`.

**Comment by MindTooth** [comment]
More:

- [ ] `Makefile`
- [ ] Documenting the new steps

**Comment by tomrtk** [comment]
> Can we with `pip-tools` perhaps remove a lot of the dependencies that are listed explicit, if they are already pulled implicitly by our hard requirements? 

as long as we do not also import it explicitly in our code, I would say we can remove it.

**Comment by MindTooth** [comment]
Added `pre-commit` check for naming tests correctly:

```
python tests naming......................................................Failed
- hook id: name-tests-test
- exit code: 1

tests/unit/core/integration/test_schema.py does not match pattern ".*_test\.py"
tests/integration/test_regression.py does not match pattern ".*_test\.py"
tests/unit/config/test_config.py does not match pattern ".*_test\.py"
tests/unit/core/unit/test_main.py does not match pattern ".*_test\.py"
tests/unit/core/unit/test_job.py does not match pattern ".*_test\.py"
tests/integration/test_tracking.py does not match pattern ".*_test\.py"
tests/unit/core/integration/test_repository.py does not match pattern ".*_test\.py"
tests/unit/core/unit/test_frame.py does not match pattern ".*_test\.py"
tests/unit/detection/test_api.py does not match pattern ".*_test\.py"
tests/unit/tracing/test_tracker.py does not match pattern ".*_test\.py"
tests/integration/test_detection.py does not match pattern ".*_test\.py"
tests/unit/core/unit/test_interface.py does not match pattern ".*_test\.py"
tests/unit/tracing/test_api.py does not match pattern ".*_test\.py"
tests/unit/core/unit/test_api_helpers.py does not match pattern ".*_test\.py"
tests/unit/tracing/unit/test_main.py does not match pattern ".*_test\.py"
tests/unit/ui/test_main.py does not match pattern ".*_test\.py"
tests/integration/test_ui.py does not match pattern ".*_test\.py"
tests/unit/tracing/test_detection.py does not match pattern ".*_test\.py"
tests/unit/ui/test_ui.py does not match pattern ".*_test\.py"
tests/unit/core/unit/test_schema.py does not match pattern ".*_test\.py"
tests/unit/core/unit/test_project.py does not match pattern ".*_test\.py"
tests/unit/core/integration/test_repository_video.py does not match pattern ".*_test\.py"
tests/unit/core/unit/test_parse.py does not match pattern ".*_test\.py"
tests/unit/ui/test_utils.py does not match pattern ".*_test\.py"
tests/integration/test_services.py does not match pattern ".*_test\.py"
tests/unit/core/unit/test_object.py does not match pattern ".*_test\.py"
tests/unit/detection/test_main.py does not match pattern ".*_test\.py"
tests/unit/core/integration/test_repository_object.py does not match pattern ".*_test\.py"
tests/unit/core/unit/test_video.py does not match pattern ".*_test\.py"
tests/unit/tracing/test_object.py does not match pattern ".*_test\.py"
tests/unit/core/unit/test_detection.py does not match pattern ".*_test\.py"
tests/unit/core/unit/test_conversion.py does not match pattern ".*_test\.py"
tests/unit/core/integration/test_api.py does not match pattern ".*_test\.py"
tests/unit/tracing/test_bbox.py does not match pattern ".*_test\.py"
tests/unit/core/unit/test_service.py does not match pattern ".*_test\.py"
tests/unit/nina/test_nina.py does not match pattern ".*_test\.py"
```

Also reordered them a bit to keep them more logically.  Stolen loosely from [Anthony Sottile](https://github.com/asottile).

Thirdly, the `flake8-bugbear` issues:
```
flake8...................................................................Failed
- hook id: flake8
- exit code: 1

src/core/api/api.py:104:31: B008 Do not perform function calls in argument defaults.  The call is performed only once at function definition time. All calls to your function will reuse the result of that definition-time function call.  If this is intended, assign the function call to a module-level variable and use that variable as a default value.
src/core/api/api.py:105:17: B008 Do not perform function calls in argument defaults.  The call is performed only once at function definition time. All calls to your function will reuse the result of that definition-time function call.  If this is intended, assign the function call to a module-level variable and use that variable as a default value.
src/core/api/api.py:110:21: B008 Do not perform function calls in argument defaults.  The call is performed only once at function definition time. All calls to your function will reuse the result of that definition-time function call.  If this is intended, assign the function call to a module-level variable and use that variable as a default value.
src/core/api/api.py:172:31: B008 Do not perform function calls in argument defaults.  The call is performed only once at function definition time. All calls to your function will reuse the result of that definition-time function call.  If this is intended, assign the function call to a module-level variable and use that variable as a default value.
src/core/api/api.py:191:31: B008 Do not perform function calls in argument defaults.  The call is performed only once at function definition time. All calls to your function will reuse the result of that definition-time function call.  If this is intended, assign the function call to a module-level variable and use that variable as a default value.
src/core/api/api.py:220:31: B008 Do not perform function calls in argument defaults.  The call is performed only once at function definition time. All calls to your function will reuse the result of that definition-time function call.  If this is intended, assign the function call to a module-level variable and use that variable as a default value.
src/core/api/api.py:221:17: B008 Do not perform function calls in argument defaults.  The call is performed only once at function definition time. All calls to your function will reuse the result of that definition-time function call.  If this is intended, assign the function call to a module-level variable and use that variable as a default value.
src/core/api/api.py:226:21: B008 Do not perform function calls in argument defaults.  The call is performed only once at function definition time. All calls to your function will reuse the result of that definition-time function call.  If this is intended, assign the function call to a module-level variable and use that variable as a default value.
src/core/api/api.py:286:31: B008 Do not perform function calls in argument defaults.  The call is performed only once at function definition time. All calls to your function will reuse the result of that definition-time function call.  If this is intended, assign the function call to a module-level variable and use that variable as a default value.
src/core/api/api.py:352:31: B008 Do not perform function calls in argument defaults.  The call is performed only once at function definition time. All calls to your function will reuse the result of that definition-time function call.  If this is intended, assign the function call to a module-level variable and use that variable as a default value.
src/core/api/api.py:390:31: B008 Do not perform function calls in argument defaults.  The call is performed only once at function definition time. All calls to your function will reuse the result of that definition-time function call.  If this is intended, assign the function call to a module-level variable and use that variable as a default value.
src/core/api/api.py:425:31: B008 Do not perform function calls in argument defaults.  The call is performed only once at function definition time. All calls to your function will reuse the result of that definition-time function call.  If this is intended, assign the function call to a module-level variable and use that variable as a default value.
src/core/api/api.py:522:22: B008 Do not perform function calls in argument defaults.  The call is performed only once at function definition time. All calls to your function will reuse the result of that definition-time function call.  If this is intended, assign the function call to a module-level variable and use that variable as a default value.
src/core/api/api.py:560:23: B008 Do not perform function calls in argument defaults.  The call is performed only once at function definition time. All calls to your function will reuse the result of that definition-time function call.  If this is intended, assign the function call to a module-level variable and use that variable as a default value.
src/core/api/api.py:561:19: B008 Do not perform function calls in argument defaults.  The call is performed only once at function definition time. All calls to your function will reuse the result of that definition-time function call.  If this is intended, assign the function call to a module-level variable and use that variable as a default value.
src/core/api/api.py:562:18: B008 Do not perform function calls in argument defaults.  The call is performed only once at function definition time. All calls to your function will reuse the result of that definition-time function call.  If this is intended, assign the function call to a module-level variable and use that variable as a default value.
src/core/api/api.py:563:19: B008 Do not perform function calls in argument defaults.  The call is performed only once at function definition time. All calls to your function will reuse the result of that definition-time function call.  If this is intended, assign the function call to a module-level variable and use that variable as a default value.
src/core/model.py:796:39: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
src/detection/api.py:98:44: B008 Do not perform function calls in argument defaults.  The call is performed only once at function definition time. All calls to your function will reuse the result of that definition-time function call.  If this is intended, assign the function call to a module-level variable and use that variable as a default value.
src/tracing/api.py:73:53: B008 Do not perform function calls in argument defaults.  The call is performed only once at function definition time. All calls to your function will reuse the result of that definition-time function call.  If this is intended, assign the function call to a module-level variable and use that variable as a default value.
src/tracing/tracker.py:274:1: B024 AbstractTracker is an abstract base class, but it has no abstract methods. Remember to use the @abstractmethod decorator, potentially in conjunction with @classmethod, @property and/or @staticmethod.
src/ui/projects/api.py:143:44: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
```

**Comment by MindTooth** [comment]
I'll split out the `pre-commit` stuff into its own pull requests.  Just wanted to test stuff out. 😄

**Comment by MindTooth** [comment]
@tomrtk If we bump Python to 3.9, we can forgo `virtualenv` and use `pyton3 -m venv venv --upgrade-deps`.  Ref.: https://docs.python.org/3.9/library/venv.html#creating-virtual-environments.

Thoughts?

**Comment by MindTooth** [comment]
> > Can we with `pip-tools` perhaps remove a lot of the dependencies that are listed explicit, if they are already pulled implicitly by our hard requirements?
> 
> as long as we do not also import it explicitly in our code, I would say we can remove it.

List of hard dependencies:

## Core
* waitress (serve UI)
* flask
* flask-paginate
* fastapi
* async-exit-stack
* async-generator
* opencv-python-headless
* sqlalchemy

## sort
* filterpy
* scipy (Pullet from `seaborn` as well)

https://github.com/abewley/sort/blob/master/requirements.txt

## yolov5
* tqdm
* psutil
* seaborn
* ipython
* torch (pulled py `torchvision`)
* torchvision
* gitpython

Guess some of these are from: https://github.com/ultralytics/yolov5/blob/master/requirements.txt

**Comment by MindTooth** [comment]
Seeing as we now just use `>=` for all dependancies.  is `pip-tools` then really necessary? 🤔   Won't `python -m buiild` just suffice?

**Comment by tomrtk** [comment]
> Seeing as we now just use `>=` for all dependancies. is `pip-tools` then really necessary? thinking Won't `python -m buiild` just suffice?

it would be good enough, but we would not have a "reproducible build" of a set of deps we know to be working, as one dep. upgrade may break something -- using `pip-tools` to produce a pinned `requirements.txt` would allow us to have a pined set of dependencies for a working/deployable version.

**Comment by tomrtk** [comment]
> If we bump Python to 3.9, we can forgo `virtualenv` and use `pyton3 -m venv venv --upgrade-deps`. Ref.: https://docs.python.org/3.9/library/venv.html#creating-virtual-environments.
> 
> Thoughts?

can drop 3.8 for me, but will your example not work as it is on 3.8 as well? 
I think of `vitrualenv` as the same as `python -m venv venv`, but a bit faster and a more convenient way of making virtual environments.

**Comment by MindTooth** [comment]
> > Seeing as we now just use `>=` for all dependancies. is `pip-tools` then really necessary? thinking Won't `python -m buiild` just suffice?
> 
> it would be good enough, but we would not have a "reproducible build" of a set of deps we know to be working, as one dep. upgrade may break something -- using `pip-tools` to produce a pinned `requirements.txt` would allow us to have a pined set of dependencies for a working/deployable version.

10-4.  I'll see what I can come up with.

> > If we bump Python to 3.9, we can forgo `virtualenv` and use `pyton3 -m venv venv --upgrade-deps`. Ref.: https://docs.python.org/3.9/library/venv.html#creating-virtual-environments.
> > Thoughts?
> 
> can drop 3.8 for me, but will your example not work as it is on 3.8 as well? I think of `vitrualenv` as the same as `python -m venv venv`, but a bit faster and a more convenient way of making virtual environments.

`--upgrade-deps` is a >= 3.9 feature.  But I'll gladly update the docs to use `virtualenv` only, so it's more clearer what we support.

**Comment by tomrtk** [comment]
> ... I'll gladly update the docs to use `virtualenv` only, so it's more clearer what we support.

ah, yes. for the docs I would maybe just drop `virtualenv`? to make not needing a extra utility to follow it, but one or the other works for me :+1: .

**Comment by MindTooth** [comment]
The doc need tweaks, but I feel we can do that at a later time.

**Review by tomrtk — APPROVED** [review]

**Pull request merged**